WDL Workflows
=============

Documentation for each workflow is provided here. Although there are many workflows that serve different functions, some of the primary workflows we use most often include:

Mostly used workflows: 
 - :doc:`demux_deplete` (on every sequencing run)
 - :doc:`classify_single` (metagenomics & host depletion)
 - :doc:`assemble_denovo` (for most viruses)
 - :doc:`assemble_refbased` (for less diverse viruses, such as those from single point source human outbreaks)
 - :doc:`augur_from_assemblies` (for nextstrain-based visualization of phylogeny)
 - :doc:`genbank` (for NCBI Genbank submission)

.. toctree::
